src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+        decimals = priceFeed.decimals();
+
+        // collateralization level
+        bufferNumerator = bufferNumerator_;
+        bufferDenominator = bufferDenominator_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for margined colleteralization colculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint maxDebt = getMaxDebt(msg.sender);
+        require(debtAmount <= maxDebt, "Insufficient collateral!");
+
+        debts[msg.sender] += debtAmount;
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Get maximum debt asset borrowable, calculated based on collateralization level
+    ///@dev Collateralization level: bufferNumerator/bufferDenominator (e.g. 8/10 -> 80%) | Think of this as margining
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
+        uint maxDebt = (availableCollateral*10**decimals / get_daiETH_Price());
+        return maxDebt;
+    }
+
+    ///@notice Users repay their debt; in debt asset terms
+    ///@dev This covers partial and full repayment scenarioes
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
+        debts[msg.sender] -= debtAmount;
+
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
+        
+        if (debts[msg.sender] > 0){
+            uint collateralRequired = getCollateralRequired(debts[msg.sender]);
+            uint spareDeposit = deposits[msg.sender] - collateralRequired;
+            require(collateralAmount < spareDeposit, "Collateral unavailable!");
+        }
+        
+        deposits[msg.sender] -= collateralAmount;
```

Comment:
```suggestion
        uint debt = debts[msg.sender]; // caching this saves an SLOAD
        uint deposit = deposits[msg.sender]; // caching this saves an SLOAD
        uint collateralRequired;
        if (debts[msg.sender] > 0){
            collateralRequired = getCollateralRequired(debt);
        }
        require(withdrawAmount <= deposit - collateralRequired, "whatever");
        
        deposit[msg.sender] = deposit - withdrawAmount;
```

Reply:
Ah i thought i was saving gas with direct referencing and not explicitly storing into memory. Any suggestions on where is a good place to get started on learning the assembly stuff that encompass SLOAD et al?

Reply:
Use the debugger.  Walk through some code. 

But tbh u can just remember any time u acess a storage slot its an sload. Cold (first) sload is 2100 gas.  Warm 2nd is 100. New local mem is 6.  

Also remember SLOAD is for 1 word (32 bytes) so if u line up your storage vars correctly u could pack 2x 128 or multiple smaller sized  into a single SLOAD